external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java

```diff
@@ -77,7 +81,8 @@ public RedisKeyValueState(String namespace, JedisCommandsInstanceContainer jedis
         this.keySerializer = keySerializer;
         this.valueSerializer = valueSerializer;
         this.jedisContainer = jedisContainer;
-        this.pendingPrepare = new ConcurrentHashMap<>();
```

Comment:
Could we somehow handle the deletes by maintaining an entry (with some extra info) within the pendingPrepare itself and avoid the extra hash sets and synchronized ? Something along the lines of maintaining a map of
`key -> {ADD or DELETE, value}`. The values to be deleted could be written to a separate Redis Set during prepare.

During `delete`, insert a DELETE entry into pendingPrepare.
During `prepare`, snapshot pendingPrepare to pendingCommit and also save the DELETE entries to a redis set.
During `commit`, delete the DELETE entries from redis map.
During `get`, if key is in pendingPrepare, pendingCommit with DELETE or the key is not found return null.